pkg/proxy/iptables/proxier.go

```diff
@@ -0,0 +1,556 @@
+			return false
+		}
+	}
+	return true
+}
+
+// How often we sync iptables
+const syncIntervalIptables = 5 * time.Second
+
+// SyncLoop runs periodic work.  This is expected to run as a goroutine or as the main loop of the app.  It does not return.
+func (proxier *Proxier) SyncLoop() {
+	t := time.NewTicker(syncIntervalIptables)
+	defer t.Stop()
+	for {
+		<-t.C
+		glog.V(6).Infof("Periodic sync")
+		if err := proxier.iptablesInit(); err != nil {
+			glog.Errorf("Failed to ensure iptables: %v", err)
+		}
+		if err := proxier.syncProxyRules(); err != nil {
+			glog.Errorf("Failed to sync iptables rules: %v", err)
+		}
+	}
+}
+
+// OnServiceUpdate tracks the active set of service proxies.
+// They will be synchronized using syncProxyRules()
+func (proxier *Proxier) OnServiceUpdate(services []api.Service) {
+	proxier.mu.Lock()
+	defer proxier.mu.Unlock()
+
+	glog.V(4).Infof("Received update notice: %+v", services)
+	activeServices := make(map[proxy.ServicePortName]bool) // use a map as a set
+
+	for i := range services {
+		service := &services[i]
+
+		// if ClusterIP is "None" or empty, skip proxying
+		if !api.IsServiceIPSet(service) {
+			glog.V(3).Infof("Skipping service %s due to portal IP = %q", types.NamespacedName{service.Namespace, service.Name}, service.Spec.ClusterIP)
+			continue
+		}
+
+		for i := range service.Spec.Ports {
+			servicePort := &service.Spec.Ports[i]
+
+			serviceName := proxy.ServicePortName{types.NamespacedName{service.Namespace, service.Name}, servicePort.Name}
+			activeServices[serviceName] = true
+			serviceIP := net.ParseIP(service.Spec.ClusterIP)
+			info, exists := proxier.serviceMap[serviceName]
+			if exists && proxier.sameConfig(info, service, servicePort) {
+				// Nothing changed.
+				continue
+			}
+			if exists {
+				//Something changed.
+				glog.V(4).Infof("Something changed for service %q: stopping it", serviceName)
+				proxier.removeServiceInternal(serviceName, info)
+			}
+
+			glog.V(1).Infof("Adding new service %q at %s:%d/%s", serviceName, serviceIP, servicePort.Port, servicePort.Protocol)
+			info = proxier.addServiceInternal(serviceName)
+			info.portal.ip = serviceIP
+			info.portal.port = servicePort.Port
+			info.portal.protocol = servicePort.Protocol
+			info.deprecatedPublicIPs = service.Spec.DeprecatedPublicIPs
+			// Deep-copy in case the service instance changes
+			info.loadBalancerStatus = *api.LoadBalancerStatusDeepCopy(&service.Status.LoadBalancer)
+			info.sessionAffinityType = service.Spec.SessionAffinity
+
+			glog.V(4).Infof("info: %+v", info)
+		}
+	}
+
+	for name, info := range proxier.serviceMap {
+		if !activeServices[name] {
+			glog.V(1).Infof("Removing service %q", name)
+			proxier.removeServiceInternal(name, info)
+		}
+	}
+
+	glog.V(6).Infof("Syncing rules.")
+	if err := proxier.iptablesInit(); err != nil {
+		glog.Errorf("Failed to ensure iptables: %v", err)
+	}
+	// this locks proxier.mu, hence the closure above
+	if err := proxier.syncProxyRules(); err != nil {
```

Comment:
syncProxyRules() takes the mutex.

You've gotten to the hard part of multi-threading - figuring out WTF just happened.